### DISCUSSION TITLE
Performance improvements

### DISCUSSION THREAD

**Discussion by ThomasHilger** [shapely/shapely#1336] [Ideas]
Hi all,

the implementation in `affine_pts` of `affine_transform` in [`affinity.py`](https://github.com/shapely/shapely/blob/f048c1816ad5782c2e70b42c55e5d74fd19ce7e7/shapely/affinity.py#L67) explicitly writes out the matrix-vector operations component by component.
I was wondering what the reason for doing this instead of utilizing numpy operations (provided the coords are compatible to numpy arrays), which would give better performance with less lines of code. Depending on the number of vertices there could be an enormous gain in performance, unless there is a reason for keeping the components explicit, which I don't see right now.

Cheers.

**Comment by jorisvandenbossche** [comment]
That's a good question. I think the main reason is that, historically, shapely didn't depend on numpy, and we only started requiring numpy recently on the main branch. So before we were even iterating over each individual coordinate pair. I ported that to vectorized operations in https://github.com/shapely/shapely/commit/4a44ed06e85cf50e63802befae2bb186f402d5a8, but didn't think of actually using numpy's matrix multiplication functionality. 

A pull request is certainly welcome!

**Comment by mwtoews** [comment]
As @jorisvandenbossche said, the function was designed without a numpy dependency, assuming coordinates were lists of tuples with 2 or 3 floats. But with shapely-2.0 currently in `main`, it can be improved as described in the question. The `coords` array should have dimensions of (N, 2) or (N, 3) for N coordinates with 2 or 3 dimensions. Would be nice to see a PR to simplify this!

**Comment by ThomasHilger** [comment]
Thx! I will happily prepare a pull request.

**Reply by mwtoews** [reply]
See #1418
